# AnnotateCADD (DNAnexus Platform App)

This is the source code for an app that runs on the DNAnexus Platform.
For more information about how to run or modify it, see
https://documentation.dnanexus.com/.

### Table of Contents

- [Introduction](#introduction)
    * [Background](#background)
    * [Dependencies](#dependencies)
        + [Docker](#docker)
        + [Resource Files](#resource-files)
- [Methodology](#methodology)
- [Running on DNANexus](#running-on-dnanexus)
    * [Inputs](#inputs)
    * [Outputs](#outputs)
    * [Command line example](#command-line-example)
        + [Batch Running](#batch-running)

## Introduction

This applet annotates vcf files generated by the [mrcepid-filterbcf](https://github.com/mrcepid-rap/mrcepid-filterbcf) applet 
with CADD variant deleteriousness scores. This applet makes heavy use of bcftools and CADD. Please see the 
[bcftools](https://samtools.github.io/bcftools/bcftools.html) or [CADD manual](https://cadd.gs.washington.edu/) for 
more information on how individual commands/tools used in this applet work.

This README makes use of DNANexus file and project naming conventions. Where applicable, an object available on the DNANexus
platform has a hash ID like:

* file – `file-1234567890ABCDEFGHIJKLMN`
* project – `project-1234567890ABCDEFGHIJKLMN`

Information about files and projects can be queried using the `dx describe` tool native to the DNANexus SDK:

```commandline
dx describe file-1234567890ABCDEFGHIJKLMN
```

**Note:** This README pertains to data included as part of the DNANexus project "MRC - Variant Filtering" (project-G2XK5zjJXk83yZ598Z7BpGPk)

### Background

CADD is one of the more reliable tools for variant interpretation. However, it has two primary limitations: 

1. Only precomputed scores are provided for all possible (3.2Gb * 3) SNVs in the genome and **NOT** most InDels
2. It requires a large amount of storage space for resource files that are used as part of its annotation model.

In particular due to the system resources required, running CADD as part of variant filtering and annotation on the 
DNANexus platform (as part of mrcepid-filterbcf) was not feasible. This is primarily due to major discrepancies in
the storage space needed by CADD (~200Gb), and the storage space needed for variant filtering (~20Gb). In brief, DNANexus
links storage space to memory and CPU requirements, so to have enough storage space for CADD resource files, we would
need a much more powerful (and expensive) AWS instance than necessary. Additionally, injesting CADD resources onto an
AWS instance takes ~1 hour. This would add significant cost to filtering. Thus, this applet runs multiple VCFs on one instance
to avoid having to waste resources on ingesting CADD annotations once per VCF file. 

### Dependencies

#### Docker

This applet uses [Docker](https://www.docker.com/) to supply dependencies to the underlying AWS instance
launched by DNANexus. The Dockerfile used to build dependencies is available as part of the MRCEpid organisation at:

https://github.com/mrcepid-rap/dockerimages/blob/main/cadd.Dockerfile

This Docker image is built off of a 20.04 Ubuntu distribution with miniconda3 pre-installed available via [dockerhub](https://hub.docker.com/r/continuumio/miniconda3).
This was done to remove issues with installing miniconda3 via Dockerfile which is a dependancy of CADD. This image is 
very light-weight and only provides basic OS installation and miniconda3. Other basic software (e.g. wget, make, and gcc) 
need to be installed manually. For more details on how to build a Docker image for use on the UKBiobank RAP, please see:

https://github.com/mrcepid-rap#docker-images

In brief, the primary **bioinformatics software** dependencies required by this Applet (and provided in the associated Docker image)
are:

* [htslib and samtools](http://www.htslib.org/)
* [bcftools](https://samtools.github.io/bcftools/bcftools.html)
* [cadd](https://cadd.gs.washington.edu/)
    * For more details on how to install CADD, please see the [CADD github repo](https://github.com/kircherlab/CADD-scripts/)

This list is not exhaustive and does not include dependencies of dependencies and software needed
to acquire other resources (e.g. wget). See the referenced Dockerfile for more information.

#### Resource Files

This app also makes use of several resource files that are specific to CADD. These files were all downloaded from the 
[CADD Downloads website](https://cadd.gs.washington.edu/download). These are provided as part of the DNANexus 
project "MRC - Variant Filtering" (project-G2XK5zjJXk83yZ598Z7BpGPk) in the folder `project_resources/` with specific
directories listed here:

* CADD resource files – `/project_resources/cadd_files/` 
* CADD known VEP hg38 cache - `/project_resources/vep_cadd_files/`

## Methodology

This applet is step 2 (mrc-annotatecadd) of the rare variant testing pipeline developed by Eugene Gardner for the UKBiobank
RAP at the MRC Epidemiology Unit:

![](https://github.com/mrcepid-rap/.github/blob/main/images/RAPPipeline.png)

This applet is fairly straightforward. Given a list of VCF files (the number of files is up to the user), this applet 
will annotate all variants in that VCF with CADD v1.6. The command line is straightforward:

```commandline
CADD-scripts/CADD.sh -g GRCh38 -o variants.cadd.tsv.gz variants.vcf
```

The difficulty was in getting CADD to run properly on AWS, which is unnecessary to document here. For more specifics on
this, please see the Dockerfile cited in [Docker](#docker) and the applet source code at `src/mrcepid-annotatecadd.py`

## Running on DNANexus

### Inputs

|input|description             |
|---- |------------------------|
|input_vcfs  | List of input vcf files from [mrcepid-filterbcf](https://github.com/mrcepid-rap/mrcepid-filterbcf) to annotate with CADD |

### Outputs

|output                 | description       |
|-----------------------|-------------------|
|output_vcfs            |  Output VCF(s) with filtered genotypes and sites, annotated with CADD         |
|output_vcf_idxs        |  .csi index files for `output_vcfs`                                           |
|output_veps            |  tabix indexed TSV file with all variants from `outputvcfs` with annotations  |
|output_vep_idxs        |  .tbi index files for `output_veps`                                           |
|output_per_samples     | .tsv file of variants per-individual to collate number of variants per person |

`output_vcfs` is a standard vcf.gz format file with INFO fields derived from VEP annotations. These fields are identical 
to those in the paired .tsv.gz file provided with the `output_veps` output. These are the following:

| INFO Field | row number in .tsv.gz | Short Description |
| ---------- | ----------------- | -----------------------|
| AF | 7 | Allele Frequency of the allele listed in `ALT` |
| F_MISSING | 8 | Proportion of missing (./.) genotypes |
| AN | 9 | Number of possible alleles ([sample size - n.missing] * 2)
| AC | 10 | Number of non-reference alleles of the allele listed in `ALT` |
| MANE | 11 | The [MANE](https://www.ncbi.nlm.nih.gov/refseq/MANE/) transcript for this variant |
| ENST | 12 | ENSEMBL transcript (e.g. ENST) for this variant |
| ENSG | 13 | ENSEMBL gene (e.g. ENSG) for this variant | 
| BIOTYPE | 14 | VEP [biotype](https://m.ensembl.org/info/genome/genebuild/biotypes.html) of the ENST. Should always be "protein_coding" |
| SYMBOL | 15 | [HGNC Gene](https://www.genenames.org/) symbol | 
| CSQ | 16 | VEP annotated [CSQ](https://m.ensembl.org/info/genome/variation/prediction/predicted_data.html) |
| gnomAD_AF | 17 | Non-Finnish European allele frequency of this variant in gnomAD exomes. 0 if not in gnomAD |
| CADD | 18 | [CADDv1.6](https://cadd.gs.washington.edu/) phred score |
| REVEL | 19 | [REVEL](https://sites.google.com/site/revelgenomics/) score for this variant if CSQ is "missense", else nan |
| SIFT | 20 | [SIFT](https://sift.bii.a-star.edu.sg/) score for this variant if CSQ is "missense", else NA |
| POLYPHEN | 21 | [POLYPHEN](http://genetics.bwh.harvard.edu/pph2/) score for this variant if CSQ is "missense", else NA |
| LOFTEE | 22 | [LOFTEE score](https://github.com/konradjk/loftee) for this variant if PARSED_CSQ is "PTV", else NA. Variants with a "HC" value are high-confidence and should be retained for testing |
| PARSED_CSQ | 23 | Eugene-determined consequence. See the README for [mrcepid-filterbcf](https://github.com/mrcepid-rap/mrcepid-filterbcf) for more information |
| MULTI | 24 | Was this variant original multiallelic? Determined based on presence of at least one ";" in the ID field |
| INDEL | 25 | Is this variant an InDel? True if len(REF) != len(ALT) | 
| MINOR | 26 | The minor allele for this variant. Will be the same as ALT if AF < 0.5 |
| MAJOR | 27 | The minor allele for this variant. Will be the same as REF if AF < 0.5 |
| MAF | 28 | The minor allele frequency for this variant. Will be the same as AF if AF < 0.5 |
| MAC | 29 | The minor allele count for this variant. Will be the same as AC if AF < 0.5 |

### Command line example

If this is your first time running this applet within a project other than "MRC - Variant Filtering", please see our
organisational documentation on how to download and build this app on the DNANexus Research Access Platform:

https://github.com/mrcepid-rap

Running this command is fairly straightforward using the DNANexus SDK toolkit. For the input vcf(s) (provided with the flag
`-iinput_vcfs`) one can use a list of file hashes of output VCFs from `mrcepid-filterbcf`:

```commandline
# Using file hash
dx run mrcepid-annotatecadd --priority low --destination filtered_vcfs/ -iinput_vcfs={file-A12345,file-B12345}
```

Brief I/O information can also be retrieved on the command line:

```commandline
dx run mrcepid-annotatecadd --help
```

Some notes here regarding execution:
1. Outputs are automatically named based on the prefix of the input vcf full path (this is regardless of if you use hash or full path). So
   the primary VCF output for the above command-line will be `ukb23156_c1_b0_v1.norm.filtered.tagged.missingness_filtered.annotated.cadd.vcf.gz`.
   All outputs will be named using a similar convention. **Note the addition** of `.cadd` compared to the output of mrcepid-filterbcf!

2. I have set a sensible (and tested) default for compute resources on DNANexus that is baked into the json used for building the app (at `dxapp.json`)
   so setting an instance type is unnecessary. This current default is for a mem1_ssd2_v2_x8 instance (8 CPUs, 16 Gb RAM, 200Gb storage).
   If necessary to adjust compute resources, one can provide a flag like `--instance-type mem1_ssd1_v2_x16`.
   
3. Note the `{}` arround the two file hashes – this is how one provides multiple files as an input array to a DNANexus applet

#### Batch Running

t.b.d. A fast way of collating multiple VCF files together as an array needs to be determined.